File: JS/cedrus/src/models/TemplatePropertyModel.ts
import { DataType } from "./DataType";
import { ITypeStamped, isTypeStamped } from "./ITypeStamped";
import { PropertyTypeModel, isPropertyTypeModel } from "./PropertyTypeModel";
import { TemplateModel, isTemplateModel } from "./TemplateModel";

export interface TemplatePropertyModel<DataTypeT extends DataType = DataType> extends ITypeStamped<"templateProperty"> {
    template: TemplateModel;
    propertyType: PropertyTypeModel<DataTypeT>;
}

export function isTemplatePropertyModel(obj: unknown): obj is TemplatePropertyModel;
export function isTemplatePropertyModel<DataTypeT extends DataType>(obj: unknown, dataType: DataTypeT): obj is TemplatePropertyModel<DataTypeT>;
export function isTemplatePropertyModel<DataTypeT extends DataType>(obj: unknown, dataType?: DataTypeT): obj is TemplatePropertyModel<DataTypeT> {
    return isTypeStamped("templateProperty", obj)
        && "template" in obj
        && isTemplateModel(obj.template)
        && "propertyType" in obj
        && isPropertyTypeModel(obj.propertyType, dataType);
}

File: NETCore/Libs/Cedrus/Controllers/V1/Models/TemplatePropertyModel.cs
using Juniper.Cedrus.Entities;

namespace Juniper.Cedrus.Controllers.V1;

public class TemplatePropertyModel
{
    public string TypeStamp => "templateProperty";
    public TemplateModel Template { get; }
    public PropertyTypeModel PropertyType { get; }

    public TemplatePropertyModel(Template? template, PropertyType? propType)
    {
        Template = new TemplateModel(template);
        PropertyType = new PropertyTypeModel(propType);
    }
}